tools/clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -1791,11 +1823,40 @@ void CodeGenFunction::EmitStoreThroughExtVectorComponentLValue(RValue Src,
       }
       Builder.CreateStore(Vec, VecDstPtr);
     } else {
-      for (unsigned i = 0; i < VecTy->getVectorNumElements(); i++) {
-        if (llvm::Constant *Elt = Elts->getAggregateElement(i)) {
-          llvm::Value *EltGEP = Builder.CreateGEP(VecDstPtr, {Zero, Elt});
-          llvm::Value *SrcElt = Builder.CreateExtractElement(SrcVal, i);
-          Builder.CreateStore(SrcElt, EltGEP);
+
+      // If the vector pointer comes from subscripting a typed rw buffer (Buffer<>, Texture*<>, etc.),
+      // insert the elements from the load.
+      //
+      // This is to avoid the final DXIL producing a load+store for each component later down the line,
+      // as there's no clean way to associate the geps+store with each other.
+      //
+      if (IsHLSubscriptOfTypedBuffer(VecDstPtr)) {
+        llvm::Value *vec = llvm::UndefValue::get(VecTy);
+
+        llvm::SmallVector<bool, 4> Stored(VecTy->getVectorNumElements());
+        for (unsigned i = 0; i < VecTy->getVectorNumElements(); i++) {
+          if (llvm::Constant *Elt = Elts->getAggregateElement(i)) {
+            llvm::Value *SrcElt = Builder.CreateExtractElement(SrcVal, i);
+            vec = Builder.CreateInsertElement(vec, SrcElt, Elt);
+            Stored[cast<llvm::ConstantInt>(Elt)->getLimitedValue()] = true;
```

Comment:
Do we actually need this extra state array? Why can't we thread the update through starting by initializing `vec = Load`

Reply:
Fixed. I was worried somehow that could lead to the loads not being removed, but tests prove otherwise.